_posts/2025-02-10-bash_scripting_an_overview.md

```diff
@@ -0,0 +1,163 @@
+
+A more robust approach is to use `trap`.  `trap` allows you to define custom actions to be taken when specific signals are received.  A common use case is to handle the `ERR` signal, which is triggered when a command exits with a non-zero status (indicating an error).
+
+Here's how to implement error handling with `trap` in the script:
+
+```bash
+#!/bin/bash
+# ... (previous comments) ...
+
+handle_error() {
+  echo "Error on line $1 with exit status $?"
+  exit 1
+}
+
+trap 'handle_error $LINENO' ERR
+
+URL="invalid_url" # Introduce an error for demonstration.
+echo "Attempting to download data from $URL"
+curl -O "$URL"
+ssconvert -S stock_data.xlsx stock_data.csv
+cut -d, -f10 stock_data.csv.0 | head
+rm stock_data*
+
+```
+
+<a href="https://www.youtube.com/watch?v=1pQ527fGhVQ&t=895.0s" target="_blank"><img src="/images/blogs/rebecca_linux_image_at_895.00.png" alt=" OK, so in this file we have the comments at the top as before." width="450"/></a>
+
+In this modified script, `handle_error` is a function that prints an error message including the line number (`$1`, passed from `$LINENO`) and the exit status (`$?`). [ $1 refers to the first argument or the first positional parameter and $? refers to the status code or the exit status of the last command that was run.](https://www.youtube.com/watch?v=1pQ527fGhVQ&t=1071s) The `trap` command associates this function with the `ERR` signal. Setting `URL` to an invalid value demonstrates the error handling in action.
+
+## Enhancing the Script: Logging
+
+Instead of printing output directly to the terminal, it's often beneficial to redirect output and errors to files. This is called logging.
+
+You can achieve this using redirection operators:
+
+*   `>` redirects standard output (stdout) to a file.
+*   `2>` redirects standard error (stderr) to a file.
+
+Here's how to modify the script execution to implement logging:
+
+```bash
+./process.sh > output.txt 2> errors.txt
+```
+
+[ So anything that would have gone to the terminal on standard out, send that to this file instead.](https://www.youtube.com/watch?v=1pQ527fGhVQ&t=1236s)
+
+This command runs `process.sh` and redirects its standard output to `output.txt` and its standard error to `errors.txt`.
+
+To encapsulate the logging within a separate script, you can use a wrapper script like this. Here is the `logger.sh` script:
+
+```bash
+#!/bin/bash
+# ... (comments) ...
+# Options:
+#   $1: Path to the script to be logged.
+
+"$1" > "${1}_output.txt" 2> "${1}_errors.txt"
+```
+<a href="https://www.youtube.com/watch?v=1pQ527fGhVQ&t=1298.28s" target="_blank"><img src="/images/blogs/rebecca_linux_image_at_1298.28.png" alt=" OK, so I've already prepared an example of logging in a script called logger." width="450"/></a>
```

Comment:
Can we add figure titles. With multiple images, it will be good to label an image with a one line description. eg Figure 1: Screenshot of `logger.sh` script.